clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -714,6 +772,19 @@ private void handleApiVersionsResponse(List<ClientResponse> responses,
         }
         NodeApiVersions nodeVersionInfo = new NodeApiVersions(apiVersionsResponse.apiVersions());
         apiVersions.update(node, nodeVersionInfo);
+        // KIP-219: enable client-side throttling for the connection to this node if the max version of
+        // ApiVersionsResponse supported by the node is >= 2.
+        if (apiVersionsResponse.apiVersion(ApiKeys.API_VERSIONS.id).maxVersion >= 2) {
+            nodesWithClientSideThrottlingEnabled.add(node);
+            if (log.isDebugEnabled()) {
+                log.debug("Client-side throttling is enabled for node {}", node);
+            }
+        } else {
+            nodesWithClientSideThrottlingEnabled.remove(node);
```

Comment:
It seem a bit unintuitive -- if a node's ApiVersionResponse's version is smaller than 2, why would the node be in `nodesWithClientSideThrottlingEnabled` in the first place?

It maybe more intuitive to remove this `else` branch and instead remove the node from `nodesWithClientSideThrottlingEnabled` in `handleConnections()`. What do you think?

Reply:
It was to cover corner cases where a broker restarted with a lower version (like rollback).

I am not entirely sure how this can be in handleConnections(), though. It looks like just initiating api version fetch. Can you elaborate?